improver/expected_value.py

```diff
@@ -30,21 +30,100 @@
 # POSSIBILITY OF SUCH DAMAGE.
 """Calculation of expected value from a probability distribution."""
 
+import numpy as np
 from iris.coords import CellMethod
 from iris.cube import Cube
 
 from improver import PostProcessingPlugin
 from improver.ensemble_copula_coupling.ensemble_copula_coupling import (
-    ConvertProbabilitiesToPercentiles,
     RebadgePercentilesAsRealizations,
 )
-from improver.metadata.probabilistic import is_percentile, is_probability
+from improver.metadata.probabilistic import (
+    find_threshold_coordinate,
+    is_percentile,
+    is_probability,
+)
 from improver.utilities.cube_manipulation import collapse_realizations
+from improver.utilities.probability_manipulation import to_threshold_inequality
 
 
 class ExpectedValue(PostProcessingPlugin):
     """Calculation of expected value from a probability distribution"""
 
+    @staticmethod
+    def integrate_over_thresholds(cube: Cube) -> Cube:
+        """
+        Calculation of expected value for threshold data by converting from
+        cumulative distribution (CDF) to probability density (PDF), then
+        integrating over the threshold dimension.
+
+        Args:
+            cube:
+                Probabilistic data with a threshold coordinate.
+
+        Returns:
+            Expected value of probability distribution. Same shape as input cube
+            but with threshold coordinate removed.
+        """
+        # make sure that the threshold direction is "below"
+        cube = to_threshold_inequality(cube, above=False)
+        # set up threshold values, these will be needed as a multiplier during integration
+        threshold_coord = find_threshold_coordinate(cube)
+        threshold_coord_idx = cube.dim_coords.index(threshold_coord)
+        thresholds = threshold_coord.points
+        # add an extra threshold below/above with zero/one probability
+        # this ensures the PDF integral covers the full CDF probability range
+        threshold_spacing = np.mean(np.diff(thresholds))
+        thresholds_expanded = np.array(
+            [
+                thresholds[0] - threshold_spacing,
+                *thresholds,
+                thresholds[-1] + threshold_spacing,
+            ]
+        )
```

Comment:
Yes, this is pretty arbitrary - it's an attempt to gracefully degrade and produce sensible output when the input data doesn't fully bound the PDF above and below (or equivalently, the threshold representation CDF starts at a value greater than zero or finishes at a value less than one).

I had a think through the effects of using the ECC bounds and I don't think it'd be sufficient in all cases.
For example, a precipitation accumulation dataset should always contain a "greater_than_or_equal_to" threshold 0.0 with probability of 1.0 (in other words, there is certain to be zero rain or some rain), or a "less than" threshold 0.0 with probability of 0.0 (in other words, there is certain to not be negative rain).
However, if something has gone wrong elsewhere in processing probabilities, that threshold probability of 1.0 could have reduced to 0.999 or 0.0 could have increased to 0.001. The ECC bounds for precipitation accumulation are 0.0 to 0.5 metres, so the approach of using those ECC bounds wouldn't extend the threshold range here.

Another approach rather than using the mean spacing between thresholds would be to repeat the first and last spacing between thresholds. Would that be an improvement, or any other suggestions?

Reply:
Maybe `min(ECC_min_bound, thresholds[0] - threshold_spacing)` (and similar for the maximum)? I think that would solve the issue you mention here, while also handling the case where `thresholds[0] - threshold_spacing`  is greater than 0.

Reply:
I've implemented this in 3ad326c3.

Reply:
Something isn't working properly with the results, I'll have to dig into this and investigate what's gone wrong - the basic unit test example has uses a full 0 to 1 CDF range, so it shouldn't vary depending on the extra thresholds.

Reply:
This is now fixed in a1444479.
The previous tests always conveniently had spacings between thresholds of exactly 1 and that poor testing didn't pick up an incorrect integration method (applying np.trapz isn't necessary, plain multiplication is equivalent to the midpoint method over the original CDF).
I've added a parametrize on the threshold_cube fixture, so that the unit tests cover a variety of threshold spacings and pick up this kind of problem.

Reply:
Changes make sense. It might be good to also add a test where the thresholds are unequally spaced (which would often be the case in practice).